frontend/src/services/featureFlags/FeatureFlagManager.ts

```diff
@@ -0,0 +1,148 @@
+import { featureFlags } from "src/constants/environments";
+import {
+  FEATURE_FLAGS_KEY,
+  getFeatureFlagsFromCookie,
+  isValidFeatureFlag,
+  parseFeatureFlagsFromString,
+  setCookie,
+} from "src/services/featureFlags/featureFlagHelpers";
+import { ServerSideSearchParams } from "src/types/searchRequestURLTypes";
+
+import { ReadonlyRequestCookies } from "next/dist/server/web/spec-extension/adapters/request-cookies";
+import { NextRequest, NextResponse } from "next/server";
+
+/**
+ * Class for reading and managing feature flags on the server.
+ *
+ * Holds default and env var based values in class, and exposes two functions:
+ *
+ * - middleware: for use in setting feature flags on cookies in next middleware
+ * - isFeatureEnabled: for checking if a feature is currently enabled, requires
+ * passing in existing request cookies
+ *
+ */
+export class FeatureFlagsManager {
+  // Define all feature flags here! You can override these in your
+  // browser, once a default is defined here.
+  private _defaultFeatureFlags = defaultFeatureFlags;
+
+  // pass in env var values at instantiation, as these won't change during runtime
+  // this supports easier integration of the class on the client side, as server side flags can be passed down
+  private _envVarFlags;
+
+  constructor(envVarFlags: FeatureFlags) {
+    this._envVarFlags = envVarFlags;
+  }
+
+  get defaultFeatureFlags(): FeatureFlags {
+    return { ...this._defaultFeatureFlags };
+  }
+
+  private get featureFlagsFromEnvironment(): FeatureFlags {
+    return { ...this._envVarFlags };
+  }
+
+  /*
+
+ * - Flags set by environment variables are the first override to default values
+    - Flags set in the /dev/feature-flags admin view will be set in the cookie
+    - Flags set in query params will result in the flag value being stored in the cookie
+    - As query param values are set in middleware on each request, query params have the highest precedence
+
+
+  */
+  get featureFlags(): FeatureFlags {
+    return {
+      ...this.defaultFeatureFlags,
+      ...this.featureFlagsFromEnvironment,
+    };
+  }
+
+  /**
+   * Check whether a feature flag is enabled
+   *
+   * @description  Server side flag values are interpreted as follows:
+   *
+   * - Default values and any environment variable overrides are grabbed from the class
+   * - Any flags from the passed in request cookie will override the class's stored flags
+   * - Any flags set in the query param are a final override, though these likely will have already been set on the cookie
+   * by the middleware
+   *
+   * @param name - Feature flag name
+   * @param cookies - server side cookies to check for enabled flags
+   * @param searchParams - search params to check for enabled flags
+   * @example isFeatureEnabled("featureFlagName")
+   */
+  isFeatureEnabled(
+    name: string,
+    cookies: NextRequest["cookies"] | ReadonlyRequestCookies,
+    searchParams?: ServerSideSearchParams,
```

Comment:
These are query params, not just for the search page AFAICT, so `queryParams` as an arg would make more sense IMO.

Reply:
not sure if you're saying that there is a better type to be using here or that we should rename the parameter. The nomenclature in the app is a bit confusing, and I confused myself enough while thinking about this that I had to look things up. https://stackoverflow.com/a/39294675 It sounds like "query params" or "search params" are not defined in any spec, but Next uses the term "search params" in their "useSearchParams" hook, and that's backed up by terminology in the URL API. "Query params" is definitely a very common phrase though. Not sure which way we'd want to go

Reply:
I meant the param name. Not sure why next uses "search" but imagine it is b/c they want to make it easier for new devs to create search pages.